src/utils/allocators.zig

```diff
@@ -514,6 +520,184 @@ pub fn RecycleFBA(config: struct {
     };
 }
 
+/// Always allocates from the backing allocator in batches at least as large as batch_size.
+pub const BatchAllocator = struct {
+    backing_allocator: Allocator,
+    batch_size: usize,
+    last_batch: Atomic(?*Batch) = Atomic(?*Batch).init(null),
+    new_batch_lock: std.Thread.RwLock = .{},
+    new_batch_waiters: Atomic(usize) = Atomic(usize).init(0),
+    new_batch_wait_lock: std.Thread.RwLock = .{},
+
+    const Self = @This();
+
+    const Batch = struct {
+        ptr: [*]u8,
+        len: usize,
+        fba: FixedBufferAllocator,
```

Comment:
why `ptr` and `len` instead of `slice: []u8`, and why separately outside of `fba`?

Reply:
I guess I did this for convenience, without putting much thought into it.

f833bf9f07dd43a0f53961eb553de57609efcb64

Reply:
Hm, I will say I didn't realize the full relationship, it's now clearer through the updated code; not pretty, but I think that's better than hiding that representation.